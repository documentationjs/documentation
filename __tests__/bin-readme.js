import path from 'path';
import { exec } from 'child_process';
import tmp from 'tmp';
import fs from 'fs-extra';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function documentation(args, options, parseJSON) {
  return new Promise((resolve, reject) => {
    if (!options.cwd) {
      options.cwd = __dirname;
    }

    options.maxBuffer = 1024 * 1024;

    args.unshift(
      'node ' + path.join(__dirname, '..', 'bin', 'documentation.js')
    );

    exec(args.join(' '), options, (err, res) => {
      resolve(res);
    });
  });
}

describe('readme autodetection of different filenames', function () {
  const fixtures = path.join(__dirname, 'fixture/readme');
  const sourceFile = path.join(fixtures, 'index.js');
  let d;
  let removeCallback;

  beforeEach(() => {
    const dirEntry = tmp.dirSync({ unsafeCleanup: true });
    d = dirEntry.name;
    fs.copySync(
      path.join(fixtures, 'README.input.md'),
      path.join(d, 'readme.markdown')
    );
    fs.copySync(path.join(fixtures, 'index.js'), path.join(d, 'index.js'));
  });

  test('updates readme.markdown', async function () {
    await documentation(['readme index.js -s API'], { cwd: d });
    const outputPath = path.join(d, 'readme.markdown');
    expect(fs.readFileSync(outputPath, 'utf-8')).toMatchSnapshot();
  });
});

describe('readme command', function () {
  const fixtures = path.join(__dirname, 'fixture/readme');
  const sourceFile = path.join(fixtures, 'index.js');
  let d;
  let removeCallback;

  beforeEach(() => {
    const dirEntry = tmp.dirSync({ unsafeCleanup: true });
    d = dirEntry.name;
    fs.copySync(
      path.join(fixtures, 'README.input.md'),
      path.join(d, 'README.md')
    );
    fs.copySync(path.join(fixtures, 'index.js'), path.join(d, 'index.js'));
  });

  // run tests after setting up temp dir

  test('--diff-only: changes needed', async function () {
    const before = fs.readFileSync(path.join(d, 'README.md'), 'utf-8');
    try {
      await documentation(['readme index.js --diff-only -s API'], {
        cwd: d
      });
    } catch (err) {
      const after = fs.readFileSync(path.join(d, 'README.md'), 'utf-8');
      expect(err).toBeTruthy();
      expect(err.code).not.toBe(0);
      expect(after).toEqual(before);
    }
  });

  test('updates README.md', async function () {
    await documentation(['readme index.js -s API'], { cwd: d });
    const outputPath = path.join(d, 'README.md');
    expect(fs.readFileSync(outputPath, 'utf-8')).toMatchSnapshot();
  });

  test('--readme-file', async function () {
    fs.copySync(
      path.join(fixtures, 'README.input.md'),
      path.join(d, 'other.md')
    );
    await documentation(['readme index.js -s API --readme-file other.md'], {
      cwd: d
    });
    const actual = fs.readFileSync(path.join(d, 'other.md'), 'utf8');
    expect(actual).toMatchSnapshot();
  });

  test('--diff-only: changes NOT needed', function () {
    fs.copySync(
      path.join(fixtures, 'README.output.md'),
      path.join(d, 'uptodate.md')
    );
    return documentation(
      ['readme index.js --diff-only -s API --readme-file uptodate.md'],
      { cwd: d }
    ).then(stdout => {
      // t.match(stdout, 'is up to date.');
    });
  });

  test('-s: not found', async function () {
    fs.copySync(
      path.join(fixtures, 'README.output.md'),
      path.join(d, 'uptodate.md')
    );
    try {
      await documentation(
        ['readme index.js --diff-only -s NOTFOUND --readme-file uptodate.md'],
        { cwd: d }
      );
    } catch (err) {
      expect(err).toBeTruthy();
    }
  });

  test('requires -s option', async function () {
    try {
      await documentation(['readme index.js'], { cwd: d });
    } catch (err) {
      expect(err).toBeTruthy();
      expect(err.code !== 0).toBeTruthy();
      expect(err.stderr.match(/Missing required argument/)).toBeTruthy();
    }
  });

  const badFixturePath = path.join(__dirname, 'fixture/bad/syntax.input');
  test('errors on invalid syntax', async function () {
    try {
      await documentation(
        ['readme ' + badFixturePath + ' -s API --parseExtension input'],
        { cwd: d }
      );
    } catch (err) {
      expect(err).toBeTruthy();
      expect(err.code !== 0).toBeTruthy();
    }
  });
});
