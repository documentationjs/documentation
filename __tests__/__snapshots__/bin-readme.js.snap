// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readme autodetection of different filenames updates readme.markdown 1`] = `
"# A title

# API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### Table of Contents

*   [foo](#foo)
    *   [Parameters](#parameters)
*   [bar](#bar)
    *   [Parameters](#parameters-1)

## foo

A function with documentation.

### Parameters

*   \`a\`  {string} blah

Returns **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** answer

## bar

A second function with docs

### Parameters

*   \`b\`  

# Another section
"
`;

exports[`readme command --readme-file 1`] = `
"# A title

# API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### Table of Contents

*   [foo](#foo)
    *   [Parameters](#parameters)
*   [bar](#bar)
    *   [Parameters](#parameters-1)

## foo

A function with documentation.

### Parameters

*   \`a\`  {string} blah

Returns **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** answer

## bar

A second function with docs

### Parameters

*   \`b\`  

# Another section
"
`;

exports[`readme command updates README.md 1`] = `
"# A title

# API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### Table of Contents

*   [foo](#foo)
    *   [Parameters](#parameters)
*   [bar](#bar)
    *   [Parameters](#parameters-1)

## foo

A function with documentation.

### Parameters

*   \`a\`  {string} blah

Returns **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** answer

## bar

A second function with docs

### Parameters

*   \`b\`  

# Another section
"
`;
